qt/widgets/common/src/FitDomain.cpp

```diff
@@ -0,0 +1,364 @@
+//   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+// SPDX - License - Identifier: GPL - 3.0 +
+#include "MantidQtWidgets/Common/FitDomain.h"
+#include "MantidQtWidgets/Common/FunctionBrowser/FunctionBrowserUtils.h"
+
+#include "MantidAPI/AnalysisDataService.h"
+#include "MantidAPI/FunctionFactory.h"
+#include "MantidAPI/MatrixWorkspace.h"
+#include "MantidKernel/Logger.h"
+
+#include <boost/algorithm/string.hpp>
+
+#include <algorithm>
+#include <stdexcept>
+
+using namespace Mantid::API;
+using namespace MantidQt::MantidWidgets;
+
+namespace {
+Mantid::Kernel::Logger g_log("FitDomain");
+
+IFunction_sptr createIFunction(std::string const &functionString) {
+  return FunctionFactory::Instance().createInitialized(functionString);
+}
+
+CompositeFunction_sptr toComposite(IFunction_sptr function) {
+  return std::dynamic_pointer_cast<CompositeFunction>(function);
+}
+
+CompositeFunction_sptr createEmptyComposite() {
+  return toComposite(createIFunction("name=CompositeFunction"));
+}
+
+std::vector<std::string> splitStringBy(std::string const &str,
+                                       std::string const &delimiter) {
+  std::vector<std::string> subStrings;
+  boost::split(subStrings, str, boost::is_any_of(delimiter));
+  subStrings.erase(std::remove_if(subStrings.begin(), subStrings.end(),
+                                  [](std::string const &subString) {
+                                    return subString.empty();
+                                  }),
+                   subStrings.cend());
+  return subStrings;
+}
+
+std::vector<std::string>
+getFunctionNamesInString(std::string const &functionString) {
+  std::vector<std::string> functionNames;
+  for (auto const &str : splitStringBy(functionString, ",();"))
+    if (str.substr(0, 5) == "name=")
+      functionNames.emplace_back(str.substr(5));
+  return functionNames;
+}
+
+bool isValueWithinConstraint(std::string const &constraint, double value) {
+  auto const limits = splitConstraintString(constraint).second;
+  return limits.first.toDouble() <= value && value <= limits.second.toDouble();
+}
+
+} // namespace
+
+namespace MantidQt {
+namespace MantidWidgets {
+
+FitDomain::FitDomain(std::string const &workspaceName,
+                     WorkspaceIndex workspaceIndex, double startX, double endX)
+    : m_workspaceName(workspaceName), m_workspaceIndex(workspaceIndex),
+      m_startX(startX), m_endX(endX), m_function(nullptr) {}
+
+bool FitDomain::setStartX(double startX) {
+  auto const validStartX = isValidStartX(startX);
+  if (validStartX)
+    m_startX = startX;
+  return validStartX;
+}
+
+bool FitDomain::setEndX(double endX) {
+  auto const validEndX = isValidEndX(endX);
+  if (validEndX)
+    m_endX = endX;
+  return validEndX;
+}
+
+void FitDomain::setFunction(Mantid::API::IFunction_sptr const &function) {
+  m_function = function;
+}
+
+Mantid::API::IFunction_sptr FitDomain::getFunction() const {
```

Comment:
Am I right in thinking that if we edit the return value from this it will not alter the stored value, because we taken a copy?

Reply:
Yes that is correct 👍

Reply:
should we call it `getFunctionCopy`? Just looking at the name I would expect a pointer, so changes would follow through

Reply:
Good idea